libbeat/publisher/queue/diskqueue/segments.go

```diff
@@ -0,0 +1,254 @@
+
+	// A list of the segments that have been completely read but have not yet
+	// been completely acknowledged, sorted by increasing segment ID. When the
+	// first entry has been completely acknowledged, it is removed from this
+	// list and appended to acked.
+	acking []*queueSegment
+
+	// A list of the segments that have been completely read and acknowledged
+	// and are ready to be deleted. When a segment is successfully deleted, it
+	// is removed from this list and discarded.
+	acked []*queueSegment
+
+	// The next sequential unused segment ID. This is what will be assigned
+	// to the next queueSegment we create.
+	nextID segmentID
+
+	// nextWriteOffset is the segment offset at which the next new frame
+	// should be written. This offset always applies to the last entry of
+	// writing[]. This is distinct from the endOffset field within a segment:
+	// endOffset tracks how much data _has_ been written to a segment, while
+	// nextWriteOffset also includes everything that is _scheduled_ to be
+	// written.
+	nextWriteOffset segmentOffset
+
+	// nextReadFrameID is the first frame ID in the current or pending
+	// read request.
+	nextReadFrameID frameID
+
+	// nextReadOffset is the segment offset corresponding to the frame
+	// nextReadFrameID. This offset always applies to the first reading
+	// segment: either reading[0], or writing[0] if reading is empty.
+	nextReadOffset segmentOffset
+}
+
+// segmentID is a unique persistent integer id assigned to each created
+// segment in ascending order.
+type segmentID uint64
+
+// segmentOffset is a byte index into the segment's data region.
+// An offset of 0 means the first byte after the segment file header.
+type segmentOffset uint64
+
+// The metadata for a single segment file.
+type queueSegment struct {
+	// A segment id is globally unique within its originating queue.
+	id segmentID
+
+	// The byte offset of the end of the segment's data region. This is
+	// updated when the segment is written to, and should always correspond
+	// to the end of a complete data frame. The total size of a segment file
+	// on disk is segmentHeaderSize + segment.endOffset.
+	endOffset segmentOffset
+
+	// The ID of the first frame that was / will be read from this segment.
+	// This field is only valid after a read request has been sent for
+	// this segment. (Currently it is only used to handle consumer ACKs,
+	// which can only happen after reading has begun on the segment.)
+	firstFrameID frameID
+
+	// The number of frames read from this segment during this session. This
+	// does not necessarily equal the number of frames in the segment, even
+	// after reading is complete, since the segment may have been partially
+	// read during a previous session.
+	//
+	// Used to count how many frames still need to be acknowledged by consumers.
+	framesRead int64
+}
+
+type segmentHeader struct {
+	version uint32
+}
+
+// Segment headers are currently just a 32-bit version.
+const segmentHeaderSize = 4
```

Comment:
unsafe.Sizeof(...)?

Reply:
`Sizeof` (as far as I know?) requires a concrete value, and returns a `uintptr`, so this ends up looking like `const segmentHeaderSize = uint64(unsafe.Sizeof(uint32(0)))`... so imo the current version is more readable, but if you know nicer syntax to accomplish the same thing I'd be happy to learn it :-)